credential/concept/query.go

```diff
@@ -0,0 +1,124 @@
+/*
+ * Nuts node
+ * Copyright (C) 2021 Nuts community
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+package concept
+
+// Query is an DB query abstraction.
+// it is created from the registry in the context of a concept.
+// It contains concept specific query arguments, that need to be resolved by the DB facade using the template registry.
+type Query interface {
+	// Concept returns the concept name.
+	Concept() string
+	// Parts returns the different TemplateQuery
+	Parts() []*TemplateQuery
+	// AddCriteria adds a criteria to the query.
+	AddCriteria(criteria Criteria)
+}
+
+type query struct {
+	concept string
+	parts   []*TemplateQuery
+}
+
+func (q query) Concept() string {
+	return q.concept
+}
+
+func (q query) Parts() []*TemplateQuery {
+	return q.parts
+}
+
+// AddTemplate adds a template to this query. It'll create a new TemplateQuery preconfigured with the template's hard coded values.
+func (q *query) AddTemplate(template *template) {
+	tq := TemplateQuery{
+		template: template,
+	}
+
+	// add all hardcoded values
+	for k, v := range template.fixedValues {
+		tq.Criteria = append(tq.Criteria, eq{
+			key:   k,
+			value: v,
+		})
+	}
+
+	q.parts = append(q.parts, &tq)
+}
+
+// Add a Criteria. The Criteria is added to each TemplateQuery
+func (q *query) AddCriteria(criteria Criteria) {
+	for _, tq := range q.parts {
+		tq.Criteria = append(tq.Criteria, criteria)
+	}
+}
+
+// TemplateQuery represents a query/template combination
+type TemplateQuery struct {
+	template *template
+	Criteria []Criteria
+}
+
+// Template returns the underlying template
+func (tq *TemplateQuery) Template() Template {
+	return tq.template
+}
+
+// VCType returns the VC type.
+func (tq *TemplateQuery) VCType() string {
+	return tq.template.VCType()
+}
+
+// Criteria abstracts different equality clauses, comparable to '=', '!=', 'between' and 'abc%' in SQL.
```

Comment:
not completely true, since it's all mapped to `leiaquery.And`..

Reply:
for now yes, but when another DB implementation is added....